![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1:
- Estudiante n°2:

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - **Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)**

<center>
<img src="https://media1.giphy.com/media/ZkIUY6LZJMwrSMQMWu/giphy.gif?cid=6c09b952msc755426bpsadn4sysofcpqczczi7x1pghd41sw&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=v" width="400" height="300">

# 📬 Entrega Parcial 3 (30% del Proyecto)

### 📪 Fecha de Entrega: 

Esta entrega cuenta con 2 partes:

- **Predicciones**: 1 de Diciembre, 2 de Diciembre, 3 de Diciembre, 4 de Diciembre
- **Informe:** 8 de Diciembre.

## 📖 Enunciado

Después de haber superado con éxito la implementación de sus flujos de trabajo con `Airflow`, el equipo **Deep Drinkers 🤖** ha demostrado que no solo sabe construir modelos, sino también integrarlos en pipelines reales listos para producción.

El equipo técnico de **SodAI Drinks 🥤** ha quedado tan satisfecho con la automatización de sus procesos que ahora les ha encomendado una nueva, y final, misión:  
- 📊 **Evaluar la calidad real de sus modelos en un entorno más desafiante** 
- 🧠 **Comunicar los hallazgos de forma clara y profesional frente a una comisión**.

<center>
<img src="https://i.giphy.com/btbUGSHh3f6eBjbDfh.webp" width="350">


## 📌 Predicciones
<center>
<img src="https://i.giphy.com/NTMxntb8rMzmbd1x97.webp" width="500" height="300">
</center>

Su *pipeline* automatizado debe demostrar que puede manejar datos reales en producción. **SodAI Drinks** 🥤 liberará *batches* de datos en fechas diferentes para evaluar si su sistema se adapta correctamente, los cuales simularán la llegada de información nueva (cada batch es una semana).

Lo que debe funcionar automáticamente:
- **Ejecución Automatizada**: Su DAG de `Airflow` debe ser capaz de procesar cada *batch* de datos de manera automática cuando esté disponible.
- **Monitoreo Continuo**: El *pipeline* debe evaluar la calidad de los datos entrantes y detectar posibles desviaciones respecto al conjunto original.
- **Decisión de Reentrenamiento**: Basándose en los análisis de *data drift* y métricas de rendimiento, el sistema debe determinar automáticamente si es necesario reentrenar el modelo.
- **Tracking Completo**: Todas las ejecuciones, métricas y decisiones deben quedar registradas en `MLFlow` para posterior análisis.
- **Generación de Predicciones**: Para cada *batch* procesado, el sistema debe generar las predicciones correspondientes y almacenarlas de manera organizada.

**IMPORTANTE: Es imperativo que guarden los resultados de su pipeline (métricas, tracking, predicciones, etc) para su análisis en la siguiente sección.**

### 📌Entregable: Competencia en CodaLab

<center>
<img src="https://i.giphy.com/NomCzPIGoXs3EIq77v.webp" width="300" height="300">
</center>

Como parte de esta entrega final, el equipo debe utilizar su pipeline entrenado para **generar predicciones en los nuevos conjuntos de datos mencionados anteriormente**.

Estas predicciones deben:

- Ser generadas directamente desde el *pipeline* previamente desarrollado.  
- Guardarse en archivos `.csv` siguiendo el formato requerido.  
- Subirse a la plataforma **CodaLab**, donde se realizará la evaluación final.

En CodaLab podrán:

- Ver el rendimiento de su modelo frente a los de otros equipos.  
- Obtener una puntuación basada en la métrica definida del proyecto.  

Cada set de predicciones debe ser publicado en CodaLab en las siguientes fechas:
- **Predicciones 1 📅** : 1 de Diciembre (Fecha de predicción: 01/01/25 - 05/01/25)
    - Note que para este set de predicciones, sólo tendrán acceso a los datos publicados en la Entrega 1.
- **Predicciones 2 📅** : 2 de Diciembre (Fecha de predicción: 06/01/25 - 12/01/25)
- **Predicciones 3 📅** : 3 de Diciembre (Fecha de predicción: 13/01/25 - 19/01/25)
- **Predicciones 4 📅** : 4 de Diciembre (Fecha de predicción: 20/01/25 - 26/01/25)

Para simular la llegada de nuevos datos, se publicarán los siguientes conjuntos de datos:
- **Batch 1 📅** : 2 de Diciembre (Con datos realizados del período 01/01/25 - 05/01/25)
- **Batch 2 📅** : 3 de Diciembre (Con datos realizados del período 06/01/25 - 12/01/25)
- **Batch 3 📅** : 4 de Diciembre (Con datos realizados del período 13/01/25 - 19/01/25)
- **Batch 4 📅** : 5 de Diciembre (Con datos realizados del período 20/01/25 - 26/01/25)
    - Note como este conjunto de datos se publica **después** de la competencia, por lo que sólo les servirá para la última evaluación requerida en el informe.

**IMPORTANTE:** Subir los resultados a tiempo en las tres fechas es **<u>OBLIGATORIO</u>** para la evaluación del desempeño final del equipo. **Por cada fecha en la que no se suban predicciones, se aplicará un descuento de 0.75 puntos (75 décimas) sobre la nota de la Entrega 3.**

### 🎁 Bonus [0.5 puntos]

Como incentivo adicional, se premiará a los **3 equipos con mejor performance en la métrica F1** con **0.5 puntos de puntaje adicional** sobre la nota de la Entrega 3.

## 📌 Informe: Performance del modelo [6.0 puntos]

En esta sección se espera que el equipo analice de manera retrospectiva el performance de su modelo y saque conclusiones en función de sus resultados. 

Para esto, tendrán que incurrir en 3 tipos de análisis:

- **Individual:** Deben evaluar el performance de su modelo tomando cada semana de forma aislada.

- **Comparativo:** Deben evaluar el performance de su modelo *a través* de las semanas, comparando el desempeño entre las semanas e identificando posibles tendencias.
    - Para esta parte se espera que generen gráficos de tendencia y tablas comparativas para apoyar su análisis.

- **Conclusiones y Aprendizajes:** Deben escribir sus principales conclusiones y aprendizajes de este proyecto. 

A lo largo de esta sección, se espera que respondan preguntas como:

1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 
2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?
3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron?  
4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?
4. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  
5. ¿Qué hiperparámetro fue el más importante para su modelo? 
5. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?
6. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?
7. ¿Qué limitaciones detectaron en su modelo o en los datos?  

**IMPORTANTE: Se espera que en sus respuestas hagan referencia a los artefactos (métricas, hiperparámetros, gráficos de interpretabilidad, etc) que su pipeline genera.**

<center>
<img src="https://media1.tenor.com/m/jyQ3SPT1htEAAAAd/i-love-this-performance-even-more-simon-cowell.gif" width="300" height="300">
</center>

### 📌 Análisis individual [3.0 puntos]

In [27]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

clientes_df = pd.read_parquet('clientes.parquet')
productos_df = pd.read_parquet('productos.parquet')

universe_size = len(clientes_df) * len(productos_df)


def load_data(batch_id):
    # Cargar datos reales
    data = pd.read_parquet(f"data/batch_t{batch_id}.parquet")
    data = data.groupby(['customer_id', 'product_id']).size().reset_index()

    preds = pd.read_csv(f"predictions/batch_t{batch_id}.csv",
                        names=['customer_id', 'product_id'])
    
    return data, preds

def get_batch_performance(data, preds):
    true_pairs = set(map(tuple, data[['customer_id', 'product_id']].values))
    print(f"cantidad real: {len(true_pairs)}")

    # Cargar predicciones
    pred_pairs = set(map(tuple, preds[['customer_id', 'product_id']].values))
    print(f"cantidad predicha: {len(pred_pairs)}")

    # === Métricas globales ===
    # TP = predicho y real
    tp = len(pred_pairs & true_pairs)
    print(f"TP: {tp}")

    # FP = predicho pero no real
    fp = len(pred_pairs - true_pairs)
    print(f"FP: {fp}")

    # FN = real pero no predicho
    fn = len(true_pairs - pred_pairs)
    print(f"FN: {fn}")

    # TN = universo total - todo lo demás
    tn = universe_size - tp - fp - fn
    print(f"TN: {tn}")

    y_true = [1]*tp + [0]*fp + [1]*fn + [0]*tn
    y_pred = [1]*tp + [1]*fp + [0]*fn + [0]*tn

    global_metrics = {
        "accuracy": round(accuracy_score(y_true, y_pred),2),
        "precision": round(precision_score(y_true, y_pred),2),
        "recall": round(recall_score(y_true, y_pred),2),
        "f1": round(f1_score(y_true, y_pred),2)
    }

    return global_metrics

def analyze_by_customer(data, preds):
    """Calcula precisión por cliente"""
    
    # Agrupar datos reales por cliente
    true_by_customer = data.groupby('customer_id')['product_id'].apply(set).to_dict()
    
    # Agrupar predicciones por cliente
    pred_by_customer = preds.groupby('customer_id')['product_id'].apply(set).to_dict()
    
    # Calcular métricas por cliente
    customer_data = []
    
    for customer_id in pred_by_customer.keys():
        pred_products = pred_by_customer.get(customer_id, set())
        true_products = true_by_customer.get(customer_id, set())
        
        tp = len(pred_products & true_products)
        fp = len(pred_products - true_products)
        
        precision = tp / (tp + fp) if (tp + fp) > 0 else 0
        
        customer_data.append({
            'customer_id': customer_id,
            'tp': tp,
            'fp': fp,
            'precision': precision,
            'total_predictions': len(pred_products)
        })
    
    customer_df = pd.DataFrame(customer_data)
    
    # Calcular X e Y
    total_customers = len(customer_df)
    customers_zero_precision = (customer_df['precision'] == 0).sum()
    customers_above_010 = (customer_df['precision'] > 0.10).sum()
    
    X = (customers_zero_precision / total_customers) * 100
    Y = (customers_above_010 / total_customers) * 100
    
    print(f"\n=== Análisis por Cliente ===")
    print(f"Total clientes con predicciones: {total_customers}")
    print(f"Clientes con precisión = 0: {customers_zero_precision} ({X:.2f}%)")
    print(f"Clientes con precisión > 0.10: {customers_above_010} ({Y:.2f}%)")
    
    return customer_df


def analyze_sample_customers(data, preds, sample_size=50):
    """Analiza una muestra de clientes para entender recomendaciones vs compras reales"""
    
    # Agrupar datos reales por cliente (compras reales)
    true_by_customer = data.groupby('customer_id')['product_id'].apply(set).to_dict()
    
    # Agrupar predicciones por cliente (recomendaciones)
    pred_by_customer = preds.groupby('customer_id')['product_id'].apply(set).to_dict()
    
    # Obtener clientes que tienen predicciones
    customers_with_preds = list(pred_by_customer.keys())
    
    # Tomar muestra aleatoria
    import random
    sample_customers = random.sample(customers_with_preds, min(sample_size, len(customers_with_preds)))
    
    sample_data = []
    for customer_id in sample_customers:
        num_recommendations = len(pred_by_customer.get(customer_id, set()))
        num_real_purchases = len(true_by_customer.get(customer_id, set()))
        
        sample_data.append({
            'customer_id': customer_id,
            'num_recommendations': num_recommendations,
            'num_real_purchases': num_real_purchases
        })
    
    sample_df = pd.DataFrame(sample_data)
    
    # Calcular X, Y, Z
    X = sample_df['num_recommendations'].min()
    Y = sample_df['num_recommendations'].max()
    Z = sample_df['num_real_purchases'].max()
    
    print(f"\n=== Análisis Muestra de {len(sample_df)} Clientes ===")
    print(f"Recomendaciones por cliente:")
    print(f"  - Mínimo: {X}")
    print(f"  - Máximo: {Y}")
    print(f"  - Promedio: {sample_df['num_recommendations'].mean():.2f}")
    print(f"\nCompras reales por cliente:")
    print(f"  - Máximo: {Z}")
    print(f"  - Promedio: {sample_df['num_real_purchases'].mean():.2f}")
    
    print(f"\nTexto completo:")
    print(f"En una muestra de {len(sample_df)} clientes se observa que reciben entre {X} y {Y} "
          f"recomendaciones, mientras que como máximo compran unos {Z} productos.")
    
    return sample_df


**Contexto:**

- El modelo creado categoriza los productos en 5 clases de probabilidad de compra: 'Very Low', 'Low', 'Medium', 'High', 'Very High'.
- Para predecir qué clientes comprarían en la semana siguiente, se seleccionó aquellas combinaciones cliente-producto clasificadas como 'High' o 'Very High' por el modelo.

#### 📌 Batch 1 (01/01/25 - 05/01/25) [0.75 puntos]

In [34]:
# código para evaluar el performance de su modelo
data, preds = load_data(1)

# Métricas globales
global_metrics = get_batch_performance(data, preds)
print(global_metrics)

# Análisis por cliente
customer_metrics = analyze_by_customer(data, preds)

# Análisis de muestra
sample_analysis = analyze_sample_customers(data, preds, sample_size=50)

cantidad real: 3264
cantidad predicha: 12755
TP: 1192
FP: 11563
FN: 2072
TN: 1508672
{'accuracy': 0.99, 'precision': 0.09, 'recall': 0.37, 'f1': 0.15}

=== Análisis por Cliente ===
Total clientes con predicciones: 1463
Clientes con precisión = 0: 916 (62.61%)
Clientes con precisión > 0.10: 462 (31.58%)

=== Análisis Muestra de 50 Clientes ===
Recomendaciones por cliente:
  - Mínimo: 1
  - Máximo: 23
  - Promedio: 7.46

Compras reales por cliente:
  - Máximo: 13
  - Promedio: 1.62

Texto completo:
En una muestra de 50 clientes se observa que reciben entre 1 y 23 recomendaciones, mientras que como máximo compran unos 13 productos.


In [30]:
# Filtramos solo Very High
preds2 = pd.read_csv(f"predictions_with_priority/batch_t{1}.csv")
preds2 = preds2[preds2['priority']=='Very High']

# Métricas globales
global_metrics2 = get_batch_performance(data, preds2)
global_metrics2

cantidad real: 3264
cantidad predicha: 7848
TP: 859
FP: 6989
FN: 2405
TN: 1513246


{'accuracy': 0.99, 'precision': 0.11, 'recall': 0.26, 'f1': 0.15}

El modelo identificó 12755 productos con prioridad alta o muy alta, mientras que la cantidad real transada fue de 3264.

Al compararlas con las compras efectivas se registraron 1192 verdaderos positivos, 11563 falsos positivos, 2072 falsos negativos y 1508672 verdaderos negativos.
Con estos valores, la cobertura alcanza alrededor de 37%, lo que indica que el sistema NO recupera la mayor parte de las compras reales. 
Además, emite casi 11 recomendaciones por cada interacción verdadera, lo que explica la baja precisión general (~0,09).

El accuracy fue muy alto, pero esto se debe a un desbalance de clases considerable, favoreciendo la clase negativa, en la cual el modelo predice bien la gran mayoría de los casos.

El análisis por cliente revela que más del 60% presenta precisión igual a cero y que menos del 40% supera 0,10. Esto muestra que el modelo funciona bien para un grupo reducido, pero para la mayoría sus aciertos son prácticamente aleatorios.

En una muestra de 50 clientes se observa que reciben entre 1 y 27 recomendaciones, mientras que como máximo compran unos 16 productos. Esta diferencia confirma que utilizar las clases "High" y "Very High" como criterio resulta demasiado permisivo y genera un exceso de sugerencias sin relevancia.

Al limitar el análisis a la clase "Very High", la precisión aumenta de 0.09 a 0.11, el F1 se mantiene, y el recall disminuye de 0,37 a 0,26. Este comportamiento indica que aplicar un umbral más exigente , si bien mejoraría ligeramente la precision, perjudicaría considerablemente el recall.

Desde una perspectiva operativa, si bien emitir 11 recomendaciones por cada compra real resulta viable, podría saturar a algunos clientes. 
Ajustar el umbral de decisión pareciera una medida efectiva en este sentido, pues disminuye los falsos positivos casi a la mitad (se realizan menos recomendaciones a clientes que no compraron). Sin embargo, aumentan los falsos negativos y disminuyen los veraderos positivos, haciéndose menos recomendaciones relevantes, lo que podría disminuir las ventas. De antemano pareciera que existe un problema ya sea en el preprocesamiento, es decir, el cómo se genera la clase "priority", y la elección del modelo (KNeighbors Classifier).

#### 📌 Batch 2 (06/01/25 - 12/01/25) [0.75 puntos]

In [36]:
# código para evaluar el performance de su modelo
data, preds = load_data(2)

# Métricas globales
global_metrics = get_batch_performance(data, preds)
print(global_metrics)

# Análisis por cliente
customer_metrics = analyze_by_customer(data, preds)

# Análisis de muestra
sample_analysis = analyze_sample_customers(data, preds, sample_size=50)

cantidad real: 5103
cantidad predicha: 12878
TP: 1766
FP: 11112
FN: 3337
TN: 1507284
{'accuracy': 0.99, 'precision': 0.14, 'recall': 0.35, 'f1': 0.2}

=== Análisis por Cliente ===
Total clientes con predicciones: 1461
Clientes con precisión = 0: 671 (45.93%)
Clientes con precisión > 0.10: 683 (46.75%)

=== Análisis Muestra de 50 Clientes ===
Recomendaciones por cliente:
  - Mínimo: 1
  - Máximo: 26
  - Promedio: 8.26

Compras reales por cliente:
  - Máximo: 14
  - Promedio: 3.54

Texto completo:
En una muestra de 50 clientes se observa que reciben entre 1 y 26 recomendaciones, mientras que como máximo compran unos 14 productos.


In [37]:
# Filtramos solo Very High
preds2 = pd.read_csv(f"predictions_with_priority/batch_t{2}.csv")
preds2 = preds2[preds2['priority']=='Very High']

# Métricas globales
global_metrics2 = get_batch_performance(data, preds2)
global_metrics2

cantidad real: 5103
cantidad predicha: 7507
TP: 1024
FP: 6483
FN: 4079
TN: 1511913


{'accuracy': 0.99, 'precision': 0.14, 'recall': 0.2, 'f1': 0.16}

El segundo batch contiene una cantidad real de 5103, mientras que el modelo predice una cantidad de 12878 transacciones. El sistema obtuvo 1766 TP, 11112 FP y 3337 FN. Con estos valores, el recall baja a 0,35, ligeramente bajo el valrr 0,37 del batch 1. La tasa de predicciones disminuye: ahora se producen alrededor de 7 sugerencias por cada caso real (la precisión aumenta a  0.14).

El porcentaje de clientes con precisión igual a 0 disminuye a 45,93%, mientras que el porcentaje de clientes con precisión mayor a 10% sube a 46.75%.
En la muestra de 50 clientes se vuelve a observar un fuerte desbalance: entre 2 y 29 recomendaciones por cliente frente a un máximo de 15 compras reales.

Al limitar la evaluación a la clase "Very High", la precisión se mantiene en 0.14, el F1 baja a 0.16, y el recall baja a 0.2. 
Al igual que en el batch 1, si bien esta medida erradica alrededor de la mitad de falsos positivos, introduce falsos negativos y disminuye verdaderos positivos, perjudicando el desempeño general.

En comparación con la semana previa, el modelo incrementa los TP gracias al mayor historial, sin un aumento en FP, indicando una mejora general. Sin embargo, el modelo aún no logra un buen desempeño pues la cantidad de FP sigue siendo muy alta. Es posible que esto se deba a los problemas que tiene normalmente KNeighbors Classifier con clases desbalanceadas.

A nivel operativo, el riesgo de saturar a los clientes disminuye, aunque sigue existiendo riesgo de saturar a clientes sensibles. Dado que la cantidad máxima de recomendaciones ha sido consistentemente mayor a la cantidad máxima comprada, una buena medida sería limitar la cantidad de productos sugeridos por cliente.

#### 📌 Batch 03 (13/01/25 - 19/01/25) [0.75 puntos]

In [ ]:
# código para evaluar el performance de su modelo
data, preds = load_data(3)

# Métricas globales
global_metrics = get_batch_performance(data, preds)
print(global_metrics)

# Análisis por cliente
customer_metrics = analyze_by_customer(data, preds)

# Análisis de muestra
sample_analysis = analyze_sample_customers(data, preds, sample_size=50)

In [ ]:
# Filtramos solo Very High
preds2 = pd.read_csv(f"predictions_with_priority/batch_t{3}.csv")
preds2 = preds2[preds2['priority']=='Very High']

# Métricas globales
global_metrics2 = get_batch_performance(data, preds2)
global_metrics2

#### 📌 Batch 04 (20/01/25 - 26/01/25) [0.75 puntos]

In [ ]:
# código para evaluar el performance de su modelo
data, preds = load_data(4)

# Métricas globales
global_metrics = get_batch_performance(data, preds)
print(global_metrics)

# Análisis por cliente
customer_metrics = analyze_by_customer(data, preds)

# Análisis de muestra
sample_analysis = analyze_sample_customers(data, preds, sample_size=50)

In [ ]:
# Filtramos solo Very High
preds2 = pd.read_csv(f"predictions_with_priority/batch_t{4}.csv")
preds2 = preds2[preds2['priority']=='Very High']

# Métricas globales
global_metrics2 = get_batch_performance(data, preds2)
global_metrics2

### 📌 Análisis comparativo [2.0 puntos]

In [ ]:
# código para realizar análisis de resultados entre batchs

### 📌 Conclusiones y Aprendizajes [1.0 puntos]

> [Escriba aquí sus conclusiones y aprendizajes]


Una gran limitante del modelo creado es que este categoriza los productos en clases según su probabilidad de compra, como se pedía en el enunciado de la entrega 1. Sin embargo, en esta entrega se evalúa el desempeño de la prediccion de combinaciones cliente-producto que tuvieron transacciones. Para lograr esta conversión se definió un umbral de corte (seleccionando las combinaciones con clases predichas 'High' y 'Very High'), lo cual introduce un sesgo en las estimaciones.
El desempeño del modelo probablemente habría sido mucho mejor si desde un principio se planteaba para predecir directamente "Compra" o "No compra" para cada combinación cliente-producto.

Otra limitación importante es la elección del modelo en si: el modelo k-means se escogió por tener el mejor F-1 general (considerando las 5 clases creadas), pero en la práctica solo nos importaba el desempeño en las clases con mayor prioridad. Este modelo también obtuvo el mejor F-1 en la clase "Very High", pero fue superado por otros como XGboost en "High". Si bien XGBoost pudo haber logrado una mejor generalización, por ser menos sensible al desbalance, lo más probable es que el problema haya sido el preprocesamiento, incluyendo una falta de creación de nuevas features.

## 📁 Estructura de entrega

Se espera que su entrega cuente con la siguiente estructura de carpetas:

```bash
entrega_3/
│
├── informe.ipynb # Jupyter notebook con sus resultados
│
├── predictions/ # Carpeta con las predicciones para cada batch de datos
│
└── data/ # Carpeta con los datos realizados (lo que realmente ocurrió)
```

Mucho éxito!

<center>
<img src="https://gifdb.com/images/high/may-the-force-be-with-you-anakin-skywalker-n2g5o0pm4h6iylsx.gif" width="400" height="300">
